### PR TITLE
Add type annotations

### LINKED ISSUE #24: Missing Type Annotations
There are missing type annotations for some functions in this library.

The `typing` module does not exist on CircuitPython devices so the import needs to be wrapped in try/except to catch the error for missing import. There is an example of how that is done here:

```python
try:
    from typing import List, Tuple
except ImportError:
    pass
```

Once imported the typing annotations for the argument type(s), and return type(s) can be added to the function signature. Here is an example of a function that has had this done already:

```python
def wrap_text_to_pixels(
    string: str, max_width: int, font=None, indent0: str = "", indent1: str = ""
) -> List[str]:
```
If you are new to Git or Github we have a guide about contributing to our projects here: https://learn.adafruit.com/contribute-to-circuitpython-with-git-and-github

There is also a guide that covers our CI utilities and how to run them locally to ensure they will pass in Github Actions here: https://learn.adafruit.com/creating-and-sharing-a-circuitpython-library/check-your-code In particular the pages: `Sharing docs on ReadTheDocs` and `Check your code with pre-commit` contain the tools to install and commands to run locally to run the checks.

If you are attempting to resolve this issue and need help, you can post a comment on this issue and tag both @foamyguy and @kattni or reach out to us on Discord: https://adafru.it/discord in the `#circuitpython-dev` channel.

The following locations are reported by mypy to be missing type annotations:

- [ ] adafruit_si7021.py:47
- [ ] adafruit_si7021.py:60
- [ ] adafruit_si7021.py:72
- [ ] adafruit_si7021.py:126
- [ ] adafruit_si7021.py:145
- [ ] adafruit_si7021.py:183

### LINKED ISSUE #31: Issues with Pypi installation and Blinka
I'm having issues getting the latest release of this module installed via Pypi. For example, `sudo pip install adafruit_circuitpython_si7021` appears to work, but there's no implementation installed.

The latest release showing on Pypi is 3.4.0, not 4.0.0. Looking at the assets in the [4.0.0 release](https://github.com/adafruit/Adafruit_CircuitPython_SI7021/releases/tag/4.0.0), it looks like everything still refers to 3.4.0. And when I check my `dist-packages` directory, the dist-info for the package is there, but no implementation files. I've cleared out my pip cache, and tried this on a couple of systems, but it definitely seems like things are broken in Pypi.

I've also confirmed that the package is fine in the Circuitpython library bundle and can be installed correctly via copying or `circup` when using a board such as the RP2040 Feather.

### PR DESCRIPTION
**Pull request by tekktrik**
Resolves #24 as well as the issue of trying to re-release an the latest package to PyPI by adding type annotations!

### CODE REVIEW

File: adafruit_si7021/__init__.py [hunk 1]
```diff
@@ -221,14 +227,14 @@ def heater_level(self):
         return self._heater_level
 
     @heater_level.setter
-    def heater_level(self, level):
+    def heater_level(self, level: int) -> bool:
```
Comment (FoamyGuy) [inline adafruit_si7021/__init__.py]: Should this be return `None`?

**Review by FoamyGuy — CHANGES_REQUESTED** [review]
Thanks @tekktrik. I found one setter that needs adjusted I think but everything else looking good to me.

File: adafruit_si7021/__init__.py (hunk 1 above)
Comment (tekktrik) [reply adafruit_si7021/__init__.py]: Ah, good catch!

**Review by FoamyGuy — CHANGES_REQUESTED** [review]
Are up up for doing the other file in this library as well? Even though it's newer than the typing issue it would be nice to close the issue once types are in for all of the files in the library.

One other thing I noticed that I think needs updated in the setup.py is the `py_modules` field needs to be updated to `packages` I believe. Otherwise PyPi won't get the correct files bundled into it if I understand correctly.

This is the line:
https://github.com/adafruit/Adafruit_CircuitPython_SI7021/blob/8878f1b5645a0d77b132887a2432005b168bd7e6/setup.py#L57

And it should be `packages` like this one from the display_text library: https://github.com/adafruit/Adafruit_CircuitPython_Display_Text/blob/144cbf95f51d007987772ef22b3d6a36ba184d53/setup.py#L55

This was really more a part of the previous PR than this one, but I think we want to update it before attempting another release anyhow so may as well do that in this PR if you are willing.

**Comment by tekktrik** [comment]
Absolutely, will do!

**Comment by tekktrik** [comment]
Shoot, I've got a cyclic dependency when I try to add type annotations.  Honestly the best way to fix this is to merge the two files back into one.  I feel bad turning this back into an `.mpy` file again and having to change everything, but that is an option.  The other is to still keep it as a folder but just merge `i2c_bits.py` into `__init__.py`, which will function exactly the same and it will keep the folder structure.  Want to get your thoughts @FoamyGuy and @kattni on best approach here.

**Comment by tekktrik** [comment]
There's also a third option to escape the typing in `i2c_bits.py`:

```python
def __get__(
        self,
        obj: "adafruit_si7021.SI7021",
        objtype: Optional[Type["adafruit_si7021.SI7021"]] = None,
    ) -> bool:
    ...
```

It's somewhat ugly and angers Pylance in VS Code, but it is valid.

**Comment by kattni** [comment]
I'm a little unclear on what the last update was. I wasn't involved, and it's a bit outside my wheelhouse, looking at the updated code now. Point is, I'm not sure I'm the person to provide feedback on this. I will say if you're putting it all in one file, do not leave it as a package, return it to a .py file, and I can update the guide to match. But whether that is the right solution or not should be left to someone who understands the situation better than me.

**Comment by tekktrik** [comment]
I would say that is probably the way to do it in terms of best practice.  I wish I had done typing with this to catch that cyclic dependency :( Would this be another major revision change?

**Comment by tekktrik** [comment]
My vote is to remerge (sorry again) but let me know your thoughts @FoamyGuy.  It also more future proofs it by not having to do this later if the implementation I based off of `adafruit_register` classes get pushed back there and removed from here.

**Comment by kattni** [comment]
Yes it would result in another major revision change. This is fine.

**Comment by FoamyGuy** [comment]
I am not up to speed on the reasoning behind factoring i2c_bits into it's own file. I'll read through the previous PR more thoroughly this weekend to try to get caught up enough to have more informed thoughts on the resolution.

At first blush though it does kind of seem to me like i2c_bits would be more appropriate inside of adafruit_register rather than in this library. So maybe migrating it over there now and adjusting this library to use it would "feed two birds with one scone"?

**Comment by kattni** [comment]
Keep in mind, Adafruit Register needs to remain small enough to fit on smaller boards along with the libraries that depend on it. That's why we opted to make Register SPI a separate library. This may not be enough of a change to cause an issue, but it's important to keep it in mind.

**Comment by tekktrik** [comment]
> I am not up to speed on the reasoning behind factoring i2c_bits into it's own file. I'll read through the previous PR more thoroughly this weekend to try to get caught up enough to have more informed thoughts on the resolution.
> 
> At first blush though it does kind of seem to me like i2c_bits would be more appropriate inside of adafruit_register rather than in this library. So maybe migrating it over there now and adjusting this library to use it would "feed two birds with one scone"?

Basically, the existing registers library assumes that a register is just a register that can be read and/or write, in that if you want to change or read a setting, you only need to know which register it is for either case.  This device strangely (to me, at least) has different "registers" depending on whether you want to read or write.  For this reason, I believe the datasheet refers to them as "commands" but in terms of the I2C protocol they are where the register should be.

For that reason, if you want to change a bit by reading the register than writing back, you can't use the existing classes because they assume the register to be the same for both operations.  This modified version tweaks `__get__()` and `__set__()` in the borrowed classes to account for this.

I genuinely have no idea how common this paradigm is, but I hope that summary helps!

**Comment by FoamyGuy** [comment]
Interesting, Thanks for the overview. That definitely does help add some clarity for me. 

I wonder if it would be possible to modify the existing i2c_bits class inside of adafruit_register to have some optional argument or something that allows it to be used in this way. (I haven't looked at the diff, and have only a vague understanding of the actual mechanisms at play though perhaps this is not as feasible as I am imagining).

Although if this is the only device that behaves this way (or is one of very few) then maybe the addition is not worth adding to adafruit_register and indeed does make more sense living inside of this library since it facilitates the different way that the device functions.

Definitely a good point and something to keep in mind by Kattni as well though. Any additions to the actual adafruit_register would need to be scrutinized to ensure they don't make it too large to work on the smallest devices since it is so widely used by many other drivers.

I am still not fully decided on the topic of keeping it in a separate file vs. moving back into a single file. I'll take a closer look into the code and recent changes this weekend. I think I lean toward separate file, and hope that there is some other way to avoid the circular imports problem. But without having looked into the specific code yet, I can't say for certain whether it's actually possible.

**Comment by tekktrik** [comment]
Sounds good!  That other option by escaping the imports with quotes does work, I tested it last night.  If you end think that to be the best way, from a functionality standpoint separate and merged are both valid!

**Comment by FoamyGuy** [comment]
@tekktrik I've had a look over all of the information here and in the previous PR as as well as the code itself.

I am in favor of the string types for now and keeping the files separated. 

In the longer term I'm wondering if we are missing a super type class that maybe would make a good type for `obj` in the `__get__` and `__set__` functions. There is very similar code in the `adafruit_register` library here: 
https://github.com/adafruit/Adafruit_CircuitPython_Register/blob/main/adafruit_register/i2c_bits.py#L54-L55

When we implements types in here I don't think we will know the type because it could be any driver class that makes use of i2c_bit or bits.

If we did ever decide to refactor this over into `adafruit_register` or anywhere else where it would be more generally used (assuming there is other hardware that communicates like this) then we'd end up with the same problem of not knowing for sure what type `obj` will be that specifically.

It looks to me like the main assumption made by the code (both in this library and register) about `obj` is that it will have a `i2c_device` field. I wonder if we should have a class type that conveys this. `I2CDriver` or something else that signifies that it will have the `i2c_device` field.

In this library it's relatively safe to assume that `obj` will always be a `SI7021` but the more generically used code from register that this is based on won't know. And even with the class in this library it's not truly restricted to only working on `SI7021` types. It just must have a `i2c_device` field, beyond that it doesn't seem like it would care what type it is to me. So theoretically if there were some other hardware that needed this different type of I2C communication the driver for it could import the `i2c_bits` class  from here and use it.

**Comment by tekktrik** [comment]
@FoamyGuy Okay!  I'll go ahead and use string-enclosed typing for the `i2c_bits.py` file.

I agree that it may be worth while to have some sort of super type class for `obj` in `__get__()` and `__set__()`.  You mention a super type class, but another option that would leave existing classes intact would be to use duck typing via `typing.Protocol` (`typing_extensions.Protocol` in Python 3.7).  We could define the following in `circuitpython_typing`:

```python
from typing_extensions import Protocol
from adafruit_bus_device import I2CDevice

class I2CDriver(Protocol):
    i2c_driver: I2CDevice
```

Then, in `adafruit_register`, we can type `__get__()` for example:

```python
def __get__(self, obj: I2CDriver, objtype = Optional[Type[I2CDriver]] = None):
    ...
```

Thoughts?

**Comment by FoamyGuy** [comment]
@tekktrik I like it! This seems like a great use of typing Protocol to me.

**Comment by tekktrik** [comment]
@FoamyGuy sorry for the delay!  String type annotations added, as well as the change to `setup.py`!

**Comment by tekktrik** [comment]
Additionally resolves #31

**Review by FoamyGuy — APPROVED** [review]
Looks good to me. Don't have hardware to test.

Thanks for more types @tekktrik

**Comment by FoamyGuy** [comment]
@tekktrik do you want to do the release on this one? or do you know how the number should get bumped? Do we need to go to 5.0.0 or is 4.1.0 okay

**Comment by tekktrik** [comment]
I think since it's going to remain as a a folder structure, I think it should be 4.1.0 was my understanding.  I can release it!

**Pull request merged**